src/Utility/Function.h

```diff
@@ -193,6 +193,16 @@ class RealFunc : public utl::SpatialFunction<Real>, public FunctionBase
     return std::vector<Real>(1,this->evaluate(X));
   }
 
+  //! \brief Evaluates first derivatives of the function.
+  virtual Vec3 gradient(const Vec3& X) const
```

Comment:
I'm just wondering, is it smart to use `Vec3` as return type here? Maybe `std::array<Real,3>` would be more accurate, since the returned object is not a point vector (spatial coordinates or their time-derivative) but the derivative of some scalar quantity w.r.t. a physical point. Yes, they will always have the same dimension and it will have no practical difference, except that Vec3 have some pre-defined operators particularly useful for point algebra which probably(?) are not relevant here.

Reply:
While I see your point it is kinda baked into the whole system. The idea is that if you take the gradient of a ScalarFunc your get something that can act as a VectorFunc,  You take the gradient of a VectorFunc you get a STensorFunc. And the gradient of a Tensor/STensor is a matrix4d (ideally a symmetric container but Tensor4 seemed not functional). So it would just complicate matters for little gain. 

It is also used in source functions, e.g. in Stokes where we have a RealFunc for the pressure and we add its gradient to the VecFunc source function. Being able to use operator+ there makes the code pretty.

Reply:
Yes, kind of expected that. Will have too many downsides to change it. Then keep Vec3 as return type here also.

Regarding Tensor4, I think it was only introduced to represent the 4th-order material tensor in the (inefficient) tensor-formulated FiniteDeformation integrand (class NonlinearElasticity). Therefore it has very little functionality.